metals/src/main/scala/scala/meta/internal/metals/debug/DebugDiscovery.scala

```diff
@@ -162,9 +172,278 @@ class DebugDiscovery(
+        (fqn, testCases.asScala.toList)
+      }
+      .toList
+  }
+
+  private def calculateMainDistances(
+      textDocument: TextDocument,
+      mains: Seq[b.ScalaMainClass],
+      classes: BuildTargetClasses.Classes,
+      position: Position,
+  ): Seq[Distance] = {
+    mains.flatMap { mainClass =>
+      val occOpt = textDocument.occurrences.find { occ =>
+        occ.role.isDefinition && (
+          classes.mainClasses.get(occ.symbol).contains(mainClass) ||
+            DebugDiscovery
+              .mainFromAnnotation(occ, textDocument)
+              .flatMap(classes.mainClasses.get(_))
+              .contains(mainClass)
+        )
+      }
+
+      val syntheticOccOpt = if (occOpt.isEmpty) {
+        val mainMethodName = mainClass.getClassName.split('.').last
+
+        val scala3MainOpt = textDocument.symbols
+          .find { symbolInfo =>
+            symbolInfo.symbol.contains(s".$mainMethodName().") &&
+            symbolInfo.annotations.exists { annotation =>
+              annotation.tpe match {
+                case tpe: scala.meta.internal.semanticdb.TypeRef =>
+                  tpe.symbol == "scala/main#"
+                case _ => false
+              }
+            }
+          }
+          .flatMap(symbolInfo =>
+            textDocument.occurrences.find(_.symbol == symbolInfo.symbol)
+          )
+
+        scala3MainOpt.orElse {
+          val expectedSymbol = mainClass.getClassName.replace(".", "/")
+          textDocument.symbols
+            .find { symbolInfo =>
+              symbolInfo.symbol.startsWith(expectedSymbol) &&
+              (symbolInfo.symbol
+                .endsWith("#") || symbolInfo.symbol.endsWith("."))
+            }
+            .flatMap(symbolInfo =>
+              textDocument.occurrences.find(_.symbol == symbolInfo.symbol)
+            )
+        }
+      } else None
+
+      (occOpt.orElse(syntheticOccOpt)).flatMap(_.range).map { range =>
+        val lineDiff = math.abs(range.startLine - position.getLine)
+        MainDistance(lineDiff, mainClass)
+      }
     }
   }
 
+  private def calculateTestCaseDistances(
+      allTestCases: List[(String, List[TestCaseEntry])],
+      position: Position,
+  ): Seq[Distance] = {
+    allTestCases.flatMap { case (suiteFqn, cases) =>
+      cases.map { entry =>
+        val range = entry.location.getRange()
+        val start = range.getStart()
+        val lineDiff = math.abs(start.getLine() - position.getLine)
+        TestCaseDistance(lineDiff, entry.name, suiteFqn)
+      }
+    }
+  }
+
+  private def calculateTestSuiteDistances(
+      textDocument: TextDocument,
+      allTestCases: List[(String, List[TestCaseEntry])],
+      position: Position,
+  ): Seq[Distance] = {
+    val testClassOccurrences = textDocument.occurrences.filter { occ =>
+      occ.role.isDefinition && (
+        allTestCases.exists { case (suiteFqn, _) =>
+          val expectedSymbol = suiteFqn.replace(".", "/")
+          occ.symbol.startsWith(expectedSymbol) &&
```

Comment:
Same thing as above, I would try to match the exact symbol, otherwise we can match inner classes etc.